packages/xstate-store/src/atom.ts

```diff
@@ -0,0 +1,68 @@
+import { toObserver } from './toObserver';
+import { Atom, Observer, Readable, ReadonlyAtom, Subscription } from './types';
+
+export function createAtom<T>(
+  getValue: (read: <U>(atom: Readable<U>) => U) => T
+): ReadonlyAtom<T>;
+export function createAtom<T>(initialValue: T): Atom<T>;
+export function createAtom<T>(
+  valueOrFn: T | ((read: <U>(atom: Readable<U>) => U) => T)
+): Atom<T> | ReadonlyAtom<T> {
+  const current = { value: undefined as T };
+  let observers: Set<Observer<T>> | undefined;
+  const subs = new Map<Atom<any>, Subscription>();
+
+  // Handle computed case
+  if (typeof valueOrFn === 'function') {
+    const getValue = valueOrFn as (read: <U>(atom: Atom<U>) => U) => T;
+    const read = (atom: Atom<any>) => {
+      const val = atom.get();
+      if (subs.get(atom)) {
+        return val;
+      }
+      const sub = atom.subscribe(recompute);
+      subs.set(atom, sub);
+      return val;
+    };
+
+    function recompute() {
+      current.value = getValue(read);
+      observers?.forEach((o) => o.next?.(current.value));
```

Comment:
This should help with some edge cases:
```suggestion
      const newValue = getValue(read);
      current.value = newValue;
      observers?.forEach((o) => o.next?.(newValue));
```